### PR TITLE
ATorch fsdp llama2 example

### PR DESCRIPTION
**Pull request by ZhongYingMatrix**
ATorch example llama2 add fsdp

### CODE REVIEW

File: atorch/examples/llama2/dataset_model.sh [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+        -f ${MODEL_NAME_OR_PATH%/}/tokenizer_config.json && \
+        -f ${MODEL_NAME_OR_PATH%/}/tokenizer.json && \
+        -f ${MODEL_NAME_OR_PATH%/}/tokenizer.model ]]; then
+  echo "$MODEL_NAME_OR_PATH not cached."
+  pushd /tmp
+  git clone https://github.com/shawwn/llama-dl.git
+  pushd llama-dl
+  sed 's/MODEL_SIZE="7B,13B,30B,65B"/MODEL_SIZE="'$MODEL_SIZE'"/g' llama.sh > llama$MODEL_SIZE.sh
+  sh llama$MODEL_SIZE.sh
+  pip install transformers, sentencepiece
```
Comment (skydoorkai) [inline atorch/examples/llama2/dataset_model.sh:18]: remove ,

File: atorch/examples/llama2/README.md [hunk 2]
```diff
@@ -1,4 +1,48 @@
-## Intro
+# Llama2 Pretrain/Finetune
+
+This document presents examples of using ATorch to pretrain or finetune the HuggingFace Llama2 model, including FSDP (ZeRO3) and 3D hybrid parallelism approaches.
+
+## FSDP
+
```
Comment (skydoorkai) [inline atorch/examples/llama2/README.md:6]: Fully Sharded Data Parallel (FSDP) is PyTorch's implementation of ZeRO3. This example uses FSDP for distributed training, and can be used with other training optimizations, such as mixed precision, gradient checkpointing, etc.

File: atorch/examples/llama2/fsdp_llama2_entry.sh [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        --lora_r 16 \
+        --lora_alpha 16 \
+        --lora_target_modules q_proj v_proj k_proj o_proj \
+        --lora_dropout 0.05 \
+    "
+fi
+
+python -m atorch.distributed.run --fault_tolerant --max_restarts=0 \
+    --nnodes="$WORLD_SIZE" \
+    --nproc_per_node=8 \
```
Comment (skydoorkai) [inline atorch/examples/llama2/fsdp_llama2_entry.sh:24]: NUM_GPUS_PER_NODE=$(nvidia-smi -L | wc -l)

8 -> $NUM_GPUS_PER_NODE

File: atorch/examples/llama2/README.md [hunk 4]
```diff
@@ -1,4 +1,48 @@
-## Intro
+# Llama2 Pretrain/Finetune
+
+This document presents examples of using ATorch to pretrain or finetune the HuggingFace Llama2 model, including FSDP (ZeRO3) and 3D hybrid parallelism approaches.
+
```
Comment (skydoorkai) [inline atorch/examples/llama2/README.md:4]: Llama2 model and wikitext dataset is used in the examples. If you have already downloaded llama2 model, set environment variable:  xxx=llama2_model_directory
If you have wikitext in your system, set environment variable:  xxx=wikitext_data_directory

If llama2 model or wikitext dataset is not present, the training script will automatically download them for you. Note that downloading may take quite some time.

File: atorch/examples/llama2/README.md [hunk 5]
```diff
@@ -337,13 +387,16 @@
...
-- starup script
-[llama2_entry.sh](llama2_entry.sh)
+```bash
+cd dlrover/atorch/examples/Llama2
+pip install -r requirements.txt
 
-- utils
-[utils.py](utils.py)
\ No newline at end of file
+# Configurable environment variable: DATASET_PATH, MODEL_NAME_OR_PATH, PIPELINE_PARALLEL_SIZE, MODEL_PARALLEL_SIZE, etc.
```
Comment (skydoorkai) [inline atorch/examples/llama2/README.md:400]: Explain how to set 3D parallel dimensions. Give an example of how to run in 8-gpu system.

Also make the example runnable. Currently, it crashes with 2x2x2 3D.

File: atorch/examples/llama2/fsdp_llama2_entry.sh [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+	LORA_OPT="
+        --peft_type lora \
+        --lora_r 16 \
+        --lora_alpha 16 \
+        --lora_target_modules q_proj v_proj k_proj o_proj \
+        --lora_dropout 0.05 \
+    "
+fi
+
+python -m atorch.distributed.run --fault_tolerant --max_restarts=0 \
```
Comment (skydoorkai) [inline atorch/examples/llama2/fsdp_llama2_entry.sh:23]: --max_restarts=0, 就不需要fault_tolerant, 可以都去掉。

**Review by skydoorkai — APPROVED** [review]
lgtm

**Review by skydoorkai — APPROVED** [review]
fine.
